KAFKA-17217: Batched acknowledge requests per node in ShareConsumeRequestManager

ShivsundarR opened the pull request:
In ShareConsumeRequestManager, currently every time we perform a `commitSync`/`commitAsync`/`acknowledgeOnClose`  we create one `AcknowledgeRequestState` for each call. But this can be optimised further as we can batch up the acknowledgements to be sent to the same node before the next `poll()` is invoked.

This will ensure that between 2 polls, the acknowledgements are accumulated in one request per node and then sent during poll, resulting in lesser RPC calls.

To achieve this, we are storing a pair of acknowledge request states for every node, the `first` value denotes the requestState for `commitAsync()` and the `second` value denotes the requestState for `commitSync()` and `acknowledgeOnClose()`. All the acknowledgements to be sent to a particular node are stored in the corresponding acknowledgeRequestState based on whether it was synchronous or asynchronous.

ShivsundarR committed:
Batched acknowledge requests per node

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -50,20 +50,21 @@
 /**
- * {@code ShareConsumeRequestManager} is responsible for generating {@link ShareFetchRequest} and
- * {@link ShareAcknowledgeRequest} to fetch and acknowledge records being delivered for a consumer
+ * {@code ShareConsumeRequestManager} is responsible for generating
```
nit: Comment formatting messed up without changing the words at all.

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManagerTest.java:
```diff
@@ -1034,7 +1209,8 @@ protected void checkDisconnects(final long currentTimeMs) {
                         long startMs = unsentRequest.timer().currentTimeMs() - unsentRequest.timer().elapsedMs();
-                        handler.onComplete(new ClientResponse(makeHeader(unsentRequest.requestBuilder().latestAllowedVersion()),
+                        handler.onComplete(new ClientResponse(
+                                makeHeader(unsentRequest.requestBuilder().latestAllowedVersion()),
```
nit: This would be neater with the `new ClientResponse(` on the new line in its entirety.

AndrewJSchofield commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManagerTest.java:
```diff
@@ -1022,9 +1194,12 @@ private List<UnsentRequest> removeUnsentRequestByNode(Node node) {
-            // any disconnects affecting requests that have already been transmitted will be handled
-            // by NetworkClient, so we just need to check whether connections for any of the unsent
-            // requests have been disconnected; if they have, then we complete the corresponding future
+            // any disconnects affecting requests that have already been transmitted will be
```
formatting

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -207,46 +212,32 @@ public void fetch(Map<TopicIdPartition, Acknowledgements> acknowledgementsMap) {
-        } else if (!acknowledgeRequestStates.isEmpty()) {
-            // Return empty result until all the acknowledgement request states are processed
+        } else if (areAnyAcknowledgementsLeft()) {
+            // Return empty result until all the acknowledgement request states are
```
formatting

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -259,11 +250,71 @@ private PollResult processAcknowledgements(long currentTimeMs) {
      *
      * @param acknowledgementsMap The acknowledgements to commit
-     * @param deadlineMs          Time until which the request will be retried if it fails with
+     * @param deadlineMs          Time until which the request will be retried if it
```
Formatting

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -449,7 +531,8 @@ private void handleShareFetchSuccess(Node fetchTarget,
                     acks.setAcknowledgeErrorCode(Errors.forCode(partitionData.acknowledgeErrorCode()));
                     Map<TopicIdPartition, Acknowledgements> acksMap = Collections.singletonMap(tip, acks);
-                    ShareAcknowledgementCommitCallbackEvent event = new ShareAcknowledgementCommitCallbackEvent(acksMap);
+                    ShareAcknowledgementCommitCallbackEvent event = new ShareAcknowledgementCommitCallbackEvent(
```
If you want to make the line shorter, it's more readable to break it before "new".

AndrewJSchofield reviewed (changes requested):
I've done a quick pass initially and I'll take a deeper look later today.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -259,11 +250,71 @@ private PollResult processAcknowledgements(long currentTimeMs) {
+                acknowledgeRequestState.handleAcknowledgeTimedOut(tip);
+            }
+            acknowledgeRequestState.incompleteAcknowledgements.clear();
+            isAsyncDone = true;
```
Should this line be conditional on `onCommitAsync` like the others?

ShivsundarR replied:
Yes, had missed it, updated now.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -207,46 +212,32 @@ public void fetch(Map<TopicIdPartition, Acknowledgements> acknowledgementsMap) {
+        for (Pair<AcknowledgeRequestState> requestStates : acknowledgeRequestStates.values()) {
+            isAsyncDone = false;
+            // For commitAsync
+            maybeBuildRequest(requestStates.getFirst(), currentTimeMs, true).ifPresent(unsentRequests::add);
```
I think the getFirst()/getSecond() could be better named getAsyncRequest()/getSyncRequest() since that's what first and second actually mean.

ShivsundarR replied:
Yeah makes sense, I have updated now.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -378,32 +452,38 @@ public CompletableFuture<Void> acknowledgeOnClose(final Map<TopicIdPartition, Ac
+                // there will be one acknowledge request state.
+                acknowledgeRequestStates.put(nodeId, new Pair<>(acknowledgeRequestStates.get(nodeId).getFirst(),
+                        new AcknowledgeRequestState(logContext,
+                                ShareConsumeRequestManager.class.getSimpleName() + ":1",
```
The name with `:3` at the end has been lost with this PR. Please restore this.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -324,44 +382,60 @@ public void commitAsync(final Map<TopicIdPartition, Acknowledgements> acknowledg
+                        log.debug("Added async acknowledge request for partition {} to node {}", tip.topicPartition(),
+                                node.id());
                         resultCount.incrementAndGet();
+                        if (acknowledgeRequestStates.get(nodeId).getFirst() == null) {
```
You should assign the result of `acknowledgeRequestStates.get(nodeId).getFirst()` to a variable so that you can avoid repeatedly evaluating this expression. I think you're doing it 3 times when once would be enough and also make the code clearer.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -378,32 +452,38 @@ public CompletableFuture<Void> acknowledgeOnClose(final Map<TopicIdPartition, Ac
-                ));
+
+                acknowledgeRequestStates.putIfAbsent(nodeId, new Pair<>(null, null));
+                // There can only be one commitSync()/close() happening at a time. So per node,
```
This is true but you could ensure that it's true in the code also.

ShivsundarR replied:
Right yeah have added a check now.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -287,22 +338,29 @@ public CompletableFuture<Map<TopicIdPartition, Acknowledgements>> commitSync(
-                ));
+                acknowledgeRequestStates.putIfAbsent(nodeId, new Pair<>(null, null));
+
+                // There can only be one commitSync()/close() happening at a time. So per node,
```
This is true but it would be simple to ensure that using a bit of defensive code. If we made an error in the code later, we might end up overwriting something we didn't expect and not notice.

ShivsundarR replied:
Yes makes sense, have added a check now.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -259,11 +250,71 @@ private PollResult processAcknowledgements(long currentTimeMs) {
+                    && isRequestStateEmpty(acknowledgeRequestStatePair.getValue().getSecond())) {
+                areAnyAcksLeft = true;
+            } else if (!closing) {
+                acknowledgeRequestStates.remove(acknowledgeRequestStatePair.getKey());
```
Shouldn't this be `iterator.remove()`?

ShivsundarR replied:
Yes, have updated the code.

AndrewJSchofield replied:
I don't see this update yet.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -259,11 +250,71 @@ private PollResult processAcknowledgements(long currentTimeMs) {
+        return areAnyAcksLeft;
+    }
+
+    private boolean isRequestStateEmpty(AcknowledgeRequestState acknowledgeRequestState) {
```
This could be clearer. I think `ars == null && ars.acksToSend.isEmpty && ars.ia.isEmpty() && ars.ifa.isEmpty()`.

AndrewJSchofield replied:
You could even make a method `boolean AcknowledgeRequestState.isEmpty()` so that this code was not intimately aware of the three maps within.

ShivsundarR replied:
Yes makes sense, I have added in a method for the same.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -259,11 +250,71 @@ private PollResult processAcknowledgements(long currentTimeMs) {
+            Map.Entry<Integer, Pair<AcknowledgeRequestState>> acknowledgeRequestStatePair = iterator.next();
+            if (isRequestStateEmpty(acknowledgeRequestStatePair.getValue().getFirst())
+                    && isRequestStateEmpty(acknowledgeRequestStatePair.getValue().getSecond())) {
+                areAnyAcksLeft = true;
```
This confuses me. It says "if there are no async or sync acks", but then it sets `areAnyAcksLeft = true`. Seems that the logic and the names of the methods could be a bit clearer.

ShivsundarR replied:
Yes, the name is quite wrong as we want to see if it is not empty actually, I have made the changes.

AndrewJSchofield reviewed (changes requested):
Thanks for the PR. Some more comments with more to follow I expect.

ShivsundarR committed:
Removed unnecessary formatting

ShivsundarR committed:
Addressed review comments

ShivsundarR commented:
Hi @AndrewJSchofield, Thanks for the review, the formatting changes had come in due to a setting by the IDE, I have changed it now. I have addressed the remaining comments and updated the PR.

ShivsundarR committed:
Minor change

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -876,4 +1020,34 @@ public void completeIfEmpty() {
+        private V asyncRequest;
+        private V syncRequest;
+
+        public Pair(V first, V second) {
```
nit: The arguments to the constructor should also be `asyncRequest` and `syncRequest`

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -876,4 +1020,34 @@ public void completeIfEmpty() {
+            this.asyncRequest = asyncRequest;
+        }
+
+        public void setSecond(V second) {
```
`setSyncRequest`?

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -215,37 +215,20 @@ public void fetch(Map<TopicIdPartition, Acknowledgements> acknowledgementsMap) {
         if (!unsentRequests.isEmpty()) {
             pollResult = new PollResult(unsentRequests);
-        } else if (!acknowledgeRequestStates.isEmpty()) {
+        } else if (areAnyAcknowledgementsLeft()) {
```
That means a side-effect of this predicate check in the else-if is to prune any remaining acknowledge request states. That seems like a significant side-effect to have which only runs conditionally. I would have make this part of this check at all.

ShivsundarR replied:
Right yeah, we do need to prune the acknowledgements here, I have renamed the method name to make it clearer.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -291,18 +325,27 @@ public CompletableFuture<Map<TopicIdPartition, Acknowledgements>> commitSync(
+                // Ensure there is no commitSync()/close() request already present as they are blocking calls
+                // and only one request can be active at a time.
+                if (acknowledgeRequestStates.get(nodeId).getSyncRequest() != null) {
+                    log.error("Attempt to call commitSync() when there is an existing sync request for node {}", node.id());
```
Probably ought to have the resulting future completing exceptionally also. Otherwise, it will look like the commitSync() completed normally when it will not have been attempted and only the log line will reveal the bug.

ShivsundarR replied:
Right yes, have updated now.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -382,22 +438,34 @@ public CompletableFuture<Void> acknowledgeOnClose(final Map<TopicIdPartition, Ac
+                // Ensure there is no commitSync()/close() request already present as they are blocking calls
+                // and only one request can be active at a time.
+                if (acknowledgeRequestStates.get(nodeId).getSyncRequest() != null) {
+                    log.error("Attempt to call close() when there is an existing sync request for node {}", node.id());
```
The result needs to be considered here. Maybe in this case it's less critical than for commitSync().

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -259,6 +242,57 @@ private PollResult processAcknowledgements(long currentTimeMs) {
+            } else {
+                if (acknowledgeRequestState.canSendRequest(currentTimeMs)) {
+                    acknowledgeRequestState.onSendAttempt(currentTimeMs);
+                    if (onCommitAsync) {
```
This part is wrong. `isAsyncDone` should only be set to true if `onCommitAsync` is true AND we have not built a request.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -259,6 +242,57 @@ private PollResult processAcknowledgements(long currentTimeMs) {
         return pollResult;
     }
 
+    private Optional<UnsentRequest> maybeBuildRequest(AcknowledgeRequestState acknowledgeRequestState, long currentTimeMs, boolean onCommitAsync) {
```
Thinking about this, I would like you to remove `isAsyncDone` entirely. If this method returns `Optional.empty()`, then it means that no request was generated. Then the caller can take an appropriate action based on its return value, rather than using this piece of shared state which is more fragile.

ShivsundarR replied:
Yeah so here, there are cases when if when we return `Optional.empty()`, the `isAsyncDone` variable is set to `true` and cases when we set to `false`. So as of now, we do need this variable, maybe if we have a separate map for `commitAsyncs` and `commitSyncs` in a future PR, then we can get rid of this variable.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -291,18 +325,27 @@ public CompletableFuture<Map<TopicIdPartition, Acknowledgements>> commitSync(
+                    log.error("Attempt to call commitSync() when there is an existing sync request for node {}", node.id());
+                } else {
+                    // There can only be one commitSync()/close() happening at a time. So per node, there will be one acknowledge request state representing commitSync() and close().
+                    acknowledgeRequestStates.put(nodeId, new Pair<>(acknowledgeRequestStates.get(nodeId).getAsyncRequest(), new AcknowledgeRequestState(logContext,
```
There's no need to construct a new Pair here. You can use `setSyncRequest()` instead.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -332,22 +378,32 @@ public void commitAsync(final Map<TopicIdPartition, Acknowledgements> acknowledg
                         resultCount.incrementAndGet();
+                        AcknowledgeRequestState asyncRequestState = acknowledgeRequestStates.get(nodeId).getAsyncRequest();
+                        if (asyncRequestState == null) {
+                            acknowledgeRequestStates.replace(nodeId, new Pair<>(new AcknowledgeRequestState(logContext,
```
There's no need to replace or construct a new Pair here. Just use `setAsyncRequest`.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -520,31 +588,38 @@ private void handleShareAcknowledgeSuccess(Node fetchTarget,
-                    }));
-
-                    acknowledgeRequestState.processingComplete();
+                    acknowledgeRequestState.retryRequest();
```
Doesn't there need to be an else clause here for the situation in which we are not retrying? That should `handleAcknowledgeErrorCode` and set the metrics I think.

ShivsundarR replied:
Yes, thanks had missed that.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -565,7 +640,7 @@ private void handleShareAcknowledgeFailure(Node fetchTarget,
+                metricsManager.recordFailedAcknowledgements(acknowledgeRequestState.getInFlightAcknowledgementsCount(tip));
                 acknowledgeRequestState.handleAcknowledgeErrorCode(tip, Errors.forException(error));
             }));
         } finally {
```
I think we ought to call `acknowledgeRequestState.onFailedAttempt` or `acknowledgeRequestState.onSuccessfulAttempt` for every single code path that handles a response to an AcknowledgeRequestState. I think that's not true at the moment but it's hard to tell because in some cases we do it in the ShareConsumeRequestManager methods, and in some cases we do it in the AcknowledgeRequestState methods. Please can you make this consistent to make it easier to validate.

ShivsundarR replied:
Right yeah, now I have moved all the calls into the `handleShareAcknowledge` methods for consistency.

AndrewJSchofield reviewed (changes requested):
Another round of review. I'm afraid that there are quite a lot of comments, but I think it can be simplified a bit more to improve readability. Thanks for the continued effort.

ShivsundarR committed:
Addressed review comments

ShivsundarR committed:
Rename function

ShivsundarR committed:
Fix spotless bugs

AndrewJSchofield reviewed (approved):
Thanks for the PR. lgtm subject to the build completing cleanly.

omkreddy commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -518,33 +609,49 @@ private void handleShareAcknowledgeSuccess(Node fetchTarget,
             if (!handler.handleResponse(response, requestVersion)) {
                 acknowledgeRequestState.onFailedAttempt(currentTimeMs);
-                if (response.error().exception() instanceof RetriableException && !closing) {
+                log.warn("OnFailAttempt1");
```
unwanted log?

omkreddy commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -84,11 +83,12 @@ public class ShareConsumeRequestManager implements RequestManager, MemberStateLi
     private final long retryBackoffMaxMs;
     private boolean closing = false;
     private final CompletableFuture<Void> closeFuture;
+    private boolean isAsyncDone = false;
```
Can we make this as local variable?

ShivsundarR replied:
Yes, I have made it as a local variable now.

omkreddy commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -749,58 +869,104 @@ UnsentRequest buildRequest(long currentTimeMs) {
+                if (incompleteAcknowledgements.isEmpty()) {
+                    acknowledgementsToSend.clear();
+                } else {
+                    incompleteAcknowledgements.clear();
```
This line may not be required.

ShivsundarR replied:
We do move the acknowledgements to the inFlightAcknowledgements right, so we clear the map that we moved it from.

ShivsundarR committed:
Addressed PR comments

omkreddy commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java:
```diff
@@ -518,33 +612,48 @@ private void handleShareAcknowledgeSuccess(Node fetchTarget,
-                            metadata.topicNames().get(topic.topicId()));
-                    if (partition.errorCode() != Errors.NONE.code()) {
-                        metricsManager.recordFailedAcknowledgements(acknowledgeRequestState.getAcknowledgementsCount(tip));
+                AtomicBoolean shouldRetry = new AtomicBoolean(false);
```
This can be primitive boolean variable?

ShivsundarR replied:
I had it made it `AtomicBoolean` as we are trying to modify it inside a lambda expression. It did not allow `boolean`.

omkreddy reviewed (approved):
@ShivsundarR Thanks for the PR. LGTM
